site/_posts/2022-06-28-remix-demo.md

```diff
@@ -0,0 +1,152 @@
+
+// Create an instance of the authenticator, pass a generic with what
+// strategies will return and will store in the session
+export let authenticator = new Authenticator<User>(sessionStorage);
+
+authenticator.use(
+    new OAuth2Strategy(
+        {
+            authorizationURL: `${process.env.AUTH_URL}/authorize`,
+            tokenURL: `${process.env.AUTH_URL}/token`,
+            clientID: process.env.CLIENT_ID || "",
+            clientSecret: process.env.CLIENT_SECRET || "",
+            callbackURL: process.env.AUTH_CALLBACK_URL || "",
+
+        },
+        async ({ accessToken, refreshToken, extraParams, profile, context }) => {
+            // This function is MANDATORY for the system to work, and would be the
+            // main cause of being redirected to the /error route
+            console.log("Verified by FusionAuth!")
+        }
+    ),
+    // this is optional, but if you setup more than one OAuth2 instance you will
+    // need to set a custom name to each one
+    "FusionAuth"
+);
+```
+
+This code leans very heavily on @sergiodxa's OAuth work for Remix, the npm packages `remix-auth` and `remix-auth-oauth2`. It will handle all the tedious boilerplate necessary to negotiate between Remix apps and an OAuth2 authentication service like FusionAuth. As you can see, we just send some secret data -- and remember, this is all sent from the server side of Remix, not from client-side JavaScript! -- which helps tell FusionAuth what app you are trying to log in from, and where you want to go afterwards.
+
+If you are not logged in, our Remix `/login` route will redirect you to the FusionAuth login -- which you can [customize easily](/docs/v1/tech/themes/) with your own CSS! -- which looks like this out of the box:
+
+{% include _image.liquid src="/assets/img/blogs/connecting-fusionauth-remix/dinesh-login.png" alt="FusionAuth login screen" class="img-fluid" figure=true %}
+
+and if you loaded the Kickstart data as directed you should now be able to log in as `dinesh@fusionauth.io` with password `password`. Any specifically login-related errors, like forgetting your password or not being registered, will be handled here.
+
+If you log in successfully, you will be redirected to `/auth/callback` which checks your login again in code that should be familiar since it's almost identical to the `/login` route:
+
+```bash
+import type { LoaderFunction, ActionFunction } from "@remix-run/node";
+import { useLoaderData } from "@remix-run/react";
+import { sessionStorage } from "~/session.server";
+import { authenticator } from "~/auth.server";
+
+export const loader: LoaderFunction = async ({request}) => {
+  await authenticator.authenticate("FusionAuth", request, {
+    successRedirect: "/dashboard",
+    failureRedirect: "/login",
+  });
+}
+```
+
+and from there dumped into `/dashboard` which is just a message saying you're logged in. If you then try the same click path again, you will get passed through to `/dashboard` without logging in!
```

Comment:
Should this be a complete sentence instead of trying to continue a sentence from before the code example?